### PR TITLE
[SERD-1445] return JSON from run template

### PR DESCRIPTION
**Pull request by arthir**
Modified run_template() so that if specified, the JSON values are returned instead of the associated file ids

### CODE REVIEW

File: R/scripts.R [hunk 1]
```diff
@@ -99,12 +102,24 @@ run_civis <- function(expr, ...) {
...
+#' run_template(id, arguments = list(arg1 = 1, arg2 = 2), JSONValue=TRUE...)
 #' }
-run_template <- function(id, arguments, ...) {
+run_template <- function(id, arguments, JSONValue=FALSE, ...) {
   job <- scripts_post_custom(id, arguments = arguments, ...)
   run <- scripts_post_custom_runs(job$id)
   await(scripts_get_custom_runs, id = job$id, run_id = run$id)
-  fetch_output_file_ids(civis_script(job$id, run$id))
+  if (JSONValue) {
+     output <- fetch_output(civis_script(job$id, run$id))
```
Comment (arthir) [inline R/scripts.R]: This already seems to have the JSON values so decided to use this instead of calling json_values_get()

File: R/scripts.R [hunk 2]
```diff
@@ -99,12 +102,24 @@ run_civis <- function(expr, ...) {
...
+     names <- lapply(output, function(o) o$name)
+     file_values <- stats::setNames(lapply(output, function(o) o$value), names)
```
Comment (arthir) [inline R/scripts.R]: Also this means that the output should mimic the output of the file_ids look like a JSON with
{
"output": ....,
"output2": .....
}

File: R/scripts.R [hunk 3]
```diff
@@ -99,12 +106,31 @@ run_civis <- function(expr, ...) {
...
+     json_output <- output[sapply(output, function(o) o$objectType=="JSONValue")]
+     if (length(json_output) > 1) {
+        cat("Error in returning JSON outputs of template run -- too many JSON outputs")
+	return()
+     }
+     if (length(json_output) == 0) {
+        cat("Error in returning JSON outputs of template run -- no JSON output")
+	return()
+     }
+     return(json_output[[1]]$value)
```
Comment (arthir) [inline R/scripts.R]: Returning something if and only if there is exactly 1 JSON output

File: R/scripts.R (hunk 2 above)
Comment (arthir) [reply R/scripts.R]: Changed so only single JSON output is returned

File: R/scripts.R (hunk 3 above)
Comment (keithing) [reply R/scripts.R]: My instinct here (if `JSONValue=`TRUE`) would be the following:

* If there is 1 JSONValue return that value
* If there is more than 1 JSONValue, return the first and warn that only the first is returned.
* If there are no JSONValues returned, warn that no value was returned and return `list()`

File: R/scripts.R [hunk 4]
```diff
@@ -99,12 +106,31 @@ run_civis <- function(expr, ...) {
...
+run_template <- function(id, arguments, JSONValue=FALSE, ...) {
   job <- scripts_post_custom(id, arguments = arguments, ...)
   run <- scripts_post_custom_runs(job$id)
   await(scripts_get_custom_runs, id = job$id, run_id = run$id)
-  fetch_output_file_ids(civis_script(job$id, run$id))
+  if (JSONValue) {
+     output <- fetch_output(civis_script(job$id, run$id))
+     json_output <- output[sapply(output, function(o) o$objectType=="JSONValue")]
+     if (length(json_output) > 1) {
+        cat("Error in returning JSON outputs of template run -- too many JSON outputs")
```
Comment (keithing) [inline R/scripts.R]: `warning` seems more appropriate than `cat` here.

File: R/scripts.R (hunk 3 above)
Comment (arthir) [reply R/scripts.R]: That makes sense as well. I wasn't positive what the best behavior was for case 2. 
In the latest commit, I changed it to the behavior described.

File: R/scripts.R (hunk 4 above)
Comment (arthir) [reply R/scripts.R]: 👍 changed to warning.

**Comment by arthir** [comment]
@keithing Fixed this and ready for re-review

**Comment by keithing** [comment]
(cc @patr1ckm, we're extending your `run_template` function)

File: R/scripts.R [hunk 5]
```diff
@@ -88,8 +88,15 @@ run_civis <- function(expr, ...) {
...
 #' @param arguments list of arguments to the script.
+#' @param JSONValue bool (default FALSE) If true, returns the 
+#'                  JSON values instead of the file_ids
 #' @param ... additional arguments to \code{scripts_post_custom}
-#' @return File ids of any run outputs are returned.
+#' @return If JSONValue is FALSE, File ids of any run outputs are returned.
+#'         If JSONValue is TRUE, JSON values of first JSON run output is returned.
+#'           If there are no JSON outputs, nothing is returned
+#'           If there are more than 1 JSON outputs, error message is printed 
+#'             and nothing is returned.
```
Comment (keithing) [inline R/scripts.R]: This is stale since your update to the behavior for > 1 JSON output

Comment (arthir) [reply R/scripts.R]: and.. i think we're there now.

File: R/scripts.R [hunk 6]
```diff
@@ -99,12 +107,33 @@ run_civis <- function(expr, ...) {
 #'
 #' # Run the template
 #' run_template(id, arguments = list(arg1 = 1, arg2 = 2), ...)
+#'
+#' # Run the template and return JSON value outputs
+#' run_template(id, arguments = list(arg1 = 1, arg2 = 2), JSONValue=TRUE...)
```
Comment (keithing) [inline R/scripts.R]: missed a comma here:

`JSONValue=TRUE, ...`

Comment (keithing) [reply R/scripts.R]: Also, for consistency this should be `JSONValue = TRUE`

Comment (arthir) [reply R/scripts.R]: fixed it.

**Comment by arthir** [comment]
> Code looks good to me. Like in the python client, could you also add a test and update the Changelog here as well?

Added tests and updated changelog

**Pull request merged**